patches/api/0417-Add-PlayerPreRespawnLocationEvent.patch

```diff
@@ -0,0 +1,80 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Cryptite <cryptite@gmail.com>
+Date: Mon, 1 May 2023 15:23:34 -0500
+Subject: [PATCH] Add PlayerPreRespawnLocationEvent
+
+
+diff --git a/src/main/java/io/papermc/paper/event/player/PlayerPreRespawnLocationEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerPreRespawnLocationEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..cf81752c60fb1cd5a649c50acea759036ba9d094
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/event/player/PlayerPreRespawnLocationEvent.java
+@@ -0,0 +1,68 @@
++package io.papermc.paper.event.player;
++
++import org.apache.commons.lang3.Validate;
++import org.bukkit.Location;
++import org.bukkit.entity.Player;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.bukkit.event.player.PlayerRespawnEvent;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Called when a respawn event tries to determine the location of a respawn. This is called before {@link PlayerRespawnEvent}
++ */
++public class PlayerPreRespawnLocationEvent extends PlayerEvent {
++    private static final HandlerList handlers = new HandlerList();
++    private final PlayerRespawnEvent.RespawnReason respawnReason;
++    private Location respawnLocation;
++
++    public PlayerPreRespawnLocationEvent(@NotNull final Player respawnPlayer, PlayerRespawnEvent.RespawnReason respawnReason) {
++        super(respawnPlayer);
++        this.respawnReason = respawnReason;
++    }
++
++    /**
++     * Gets the current respawn location
++     *
++     * @return Location current respawn location
++     */
++    @Nullable
++    public Location getRespawnLocation() {
++        return this.respawnLocation;
++    }
++
++    /**
++     * Sets the new respawn location.
++     *
++     * @param respawnLocation The exact location for the respawn.<br/>
++     * <b>Note:</b> If this is provided, <b>no vanilla logic that calculates "safe" respawn locations will be done.</b> It is up to you to ensure you are
++     * providing a good respawn location for a Player.
++     */
++    public void setRespawnLocation(@NotNull Location respawnLocation) {
++        Validate.notNull(respawnLocation, "Respawn location can not be null");
```

Comment:
Should setting null be allowed as a way for a plugin to "undo" a previous listener?

Reply:
Fair point